source/fuzz/fuzzer_pass_add_composite_inserts.cpp

```diff
@@ -0,0 +1,256 @@
+                   instruction_descriptor.target_instruction_opcode() &&
+               "The opcode of the instruction we might insert before must be "
+               "the same as the opcode in the descriptor for the instruction");
+
+        // Randomly decide whether to try adding an OpCompositeInsert
+        // instruction.
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfAddingCompositeInsert())) {
+          return;
+        }
+
+        // It must be valid to insert an OpCompositeInsert instruction
+        // before |instruction_iterator|.
+        if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(
+                SpvOpCompositeInsert, instruction_iterator)) {
+          return;
+        }
+
+        // Look for available values that have composite type.
+        std::vector<opt::Instruction*> available_composites =
+            FindAvailableInstructions(
+                function, block, instruction_iterator,
+                [instruction_descriptor](
+                    opt::IRContext* ir_context,
+                    opt::Instruction* instruction) -> bool {
+                  auto instruction_type = ir_context->get_type_mgr()->GetType(
+                      instruction->type_id());
+                  if (!fuzzerutil::IsCompositeType(instruction_type)) {
+                    return false;
+                  }
+                  // No components of the composite can have type
+                  // OpTypeRuntimeArray.
+                  if (ContainsRuntimeArray(instruction_type)) {
+                    return false;
+                  }
+                  return fuzzerutil::IdIsAvailableBeforeInstruction(
+                      ir_context,
+                      FindInstruction(instruction_descriptor, ir_context),
+                      instruction->result_id());
+                });
+
+        // If there are no available values, then return.
+        if (available_composites.empty()) {
+          return;
+        }
+
+        // Choose randomly one available composite value.
+        auto available_composite =
+            available_composites[GetFuzzerContext()->RandomIndex(
+                available_composites)];
+
+        // Take a random component of the chosen composite value. If the chosen
+        // component is itself a composite, then randomly decide whether to take
+        // its component and repeat. Use OpCompositeExtract to get the
+        // component.
+        bool reached_end_node = false;
+        uint32_t current_node_type_id = available_composite->type_id();
+        uint32_t one_selected_index;
+        uint32_t num_of_components;
+        std::vector<uint32_t> index_to_replace;
+        while (!reached_end_node) {
+          num_of_components =
+              GetNumberOfComponents(GetIRContext(), current_node_type_id);
+          one_selected_index = GetFuzzerContext()->GetRandomIndexForAccessChain(
+              num_of_components);
```

Comment:
I was thinking about solving this problem by checking in lambda if length is ==0 and return false if so. I don't think that according to spec you can perform an OpCompositeInsert on an empty composite, because at least one index literal must be defined.

Reply:
Sorry for the confusion, I meant that in the following case:
```glsl
struct S {
  struct F {
  } a;
};
```
when you reach `a`, you can insert a component of type `F` into `S.a`. That being said, you can do:
```
%f = OpConstantComposite %S_F
%f1 = OpCopyObject %S_F %f
%s = OpConstantComposite %S %f
%s1 = OpCompositeInsert %S %f1 %s 0
```

Although, returning false is fine with me too.